service-commands/src/setup.js

```diff
@@ -640,67 +642,82 @@ const allUp = async ({
     ipfsAndContractsCommands.push([Service.SOLANA_PROGRAMS, SetupCommand.UP])
   }
 
-  let creatorNodesCommands = _.range(1, numCreatorNodes + 1).map(
-    serviceNumber => {
-      return [
-        [
-          Service.CREATOR_NODE,
-          SetupCommand.UPDATE_DELEGATE_WALLET,
-          { serviceNumber, ...options }
-        ],
-        [
-          Service.CREATOR_NODE,
-          SetupCommand.UP,
-          { serviceNumber, ...options, waitSec: 10 }
-        ],
-        [
-          Service.CREATOR_NODE,
-          SetupCommand.HEALTH_CHECK_RETRY,
-          { serviceNumber, ...options }
-        ],
-        [
-          Service.CREATOR_NODE,
-          SetupCommand.REGISTER,
-          { serviceNumber, ...options }
-        ]
+  const nodeUpCommands = []
+  const nodeHealthCheckCommands = []
+  const nodeRegisterCommands = []
+  // Add discovery node commands
+  for (
+    let serviceNumber = 1;
+    serviceNumber < numDiscoveryNodes + 1;
+    serviceNumber++
+  ) {
+    nodeUpCommands.push([
+      [
+        Service.DISCOVERY_PROVIDER,
+        SetupCommand.UP,
+        { serviceNumber, ...options }
       ]
-    }
-  )
-
-  let discoveryNodesCommands = _.range(1, numDiscoveryNodes + 1).map(
-    serviceNumber => {
-      return [
-        [
-          Service.DISCOVERY_PROVIDER,
-          SetupCommand.UP,
-          { serviceNumber, ...options }
-        ],
-        [
-          Service.DISCOVERY_PROVIDER,
-          SetupCommand.HEALTH_CHECK_RETRY,
-          { serviceNumber, ...options }
-        ],
-        [
-          Service.DISCOVERY_PROVIDER,
-          SetupCommand.REGISTER,
-          { retries: 2, serviceNumber, ...options }
-        ]
+    ])
+    nodeHealthCheckCommands.push([
+      [
+        Service.DISCOVERY_PROVIDER,
+        SetupCommand.HEALTH_CHECK_RETRY,
+        { serviceNumber, ...options }
       ]
-    }
-  )
+    ])
+    nodeRegisterCommands.push([
+      [
+        Service.DISCOVERY_PROVIDER,
+        SetupCommand.REGISTER,
+        { retries: 2, serviceNumber, ...options }
+      ]
+    ])
+  }
+  // Add creator node commands
+  for (
+    let serviceNumber = 1;
+    serviceNumber < numCreatorNodes + 1;
+    serviceNumber++
+  ) {
+    nodeUpCommands.push([
+      [
+        Service.CREATOR_NODE,
+        SetupCommand.UPDATE_DELEGATE_WALLET,
```

Comment:
so with the way that shellenv# works for these...i'm not sure we can parallelize until the templating logic is removed? @jowlee and @piazzatron have context too

Reply:
Hmm that's too bad - but the good news is even the non-parallel path is _much_ faster!

Reply:
@hareeshnagaraj @jowlee - I actually think this isn't true!

Every time we exec a command, we actually do so in a _new shell_:
https://github.com/AudiusProject/audius-protocol/blob/db85bcce980cf7c87b8fcb55232bf15106479d55/service-commands/src/setup.js#L61

These shells _don't_ modify the environment of the parent shell. This can be tested easily by adding the following to `service-commands.json`:

```
  "test_env": {
    "up": [
      "export TEST_1=\"Test 1\"",
      "echo $TEST_1 > $PROTOCOL_DIR/service-commands/test_env.log",
      "export TEST_2=\"Test 2\"; echo $TEST_2 >> $PROTOCOL_DIR/service-commands/test_env.log"
    ]
  }
  ```
 ```
❯ cat $PROTOCOL_DIR/service-commands/test_env.log

Test 2
```

This explains why I never ran into any errors running the stack with `--parallel`!

Reply:
oh that's huge! thanks for the clarification 🙏